source/FormationPattern.cpp

```diff
@@ -0,0 +1,102 @@
+/* FormationPattern.cpp
+Copyright (c) 2019-2022 by Peter van der Meer
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "FormationPattern.h"
+
+#include "Angle.h"
+#include "DataNode.h"
+
+#include <cmath>
+
+using namespace std;
+
+
+
+namespace {
+	const Point defaultFormationPoint = Point();
+}
+
+
+
+FormationPattern::PositionIterator::PositionIterator(const FormationPattern &pattern)
+	: pattern(pattern)
+{
+	positionIt = pattern.positions.begin();
+}
+
+
+
+const Point &FormationPattern::PositionIterator::operator*()
+{
+	if(positionIt == pattern.positions.end())
+		return defaultFormationPoint;
+	return *positionIt;
+}
+
+
+
+FormationPattern::PositionIterator &FormationPattern::PositionIterator::operator++()
+{
+	if(positionIt != pattern.positions.end())
+		++positionIt;
```

Comment:
There is smth weird about this code to me; there is no way to go back to the start of the formation?

Reply:
It is an iterator. If you want to restart, then throw away the current iterator and create a new one with FormationPattern::begin().

We could extend the iterator with a reset function if we like, but it is not strictly required.

Reply:
Iterators go ++ and -- from my understanding of it but as this has no practical applications yet it doesnt matter much I guess?

Reply:
Most iterators are "Random Access" or "Bidirectional" and indeed go in 2 directions. But i also found single direction input iterators in [a nice overview table](https://cplusplus.com/reference/iterator/) and those input iterators seemed sufficient for calculating the sequence of positions that was required.

When implementing other formations code, then we might need to implement something more complex (depending on other design choices that we make), but for now this code seems to suffice for testing of the positions code (and that is all we do in this PR).